Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp

```diff
@@ -0,0 +1,142 @@
+            auto cerr = myProcess.readAllStandardError();
+            outcome.m_logStandardOutput = AZStd::string(std.data(), std.size());
+            outcome.m_logErrorOutput = AZStd::string(cerr.data(), cerr.size());
+        }
+        outcome.m_returnCode = myProcess.exitCode();
+        return outcome;
+    }
+
+    AZStd::string WhichXacro()
+    {
+        const QString program{ "which" };
+        QStringList arguments;
+        arguments << kXacroExecutable;
+        QProcess myProcess;
+        myProcess.start(program, arguments);
+        if (myProcess.waitForFinished() && myProcess.exitStatus() == QProcess::NormalExit && myProcess.exitCode() == 0)
+        {
+            auto output = myProcess.readAllStandardOutput().trimmed();
+            return AZStd::string{ output.data(), static_cast<AZStd::string::size_type>(output.size()) };
+        }
+        return AZStd::string{};
+    }
+
+    AZStd::unordered_map<AZStd::string, AZStd::string> GetParameterFromXacroData(AZStd::vector<char> data)
+    {
+        const char* kArgName = "xacro:arg";
+        const char* kName = "name";
+        const char* kDefault = "default";
+
+        AZStd::unordered_map<AZStd::string, AZStd::string> params;
+        AZ::rapidxml::xml_document<char> doc;
+        doc.parse<AZ::rapidxml::parse_full>(data.data());
+        AZ::rapidxml::xml_node<char>* xmlRootNode = doc.first_node("robot");
+        for (AZ::rapidxml::xml_node<>* child_node = xmlRootNode->first_node(); child_node; child_node = child_node->next_sibling())
+        {
+            if (strcmp(kArgName, child_node->name()) == 0)
+            {
+                const auto* attributeName = child_node->first_attribute(kName);
+                const auto* attributeDefault = child_node->first_attribute(kDefault);
+
+                if (attributeName)
+                {
+                    AZStd::string name{ attributeName->value() };
+                    AZStd::string value;
+                    if (attributeDefault)
+                    {
+                        value = AZStd::string{ attributeDefault->value() };
+                    }
+                    params.emplace(AZStd::make_pair(AZStd::move(name), AZStd::move(value)));
+                }
+            }
+        }
+        return params;
+    }
+
+    AZStd::unordered_map<AZStd::string, AZStd::string> GetParameterFromXacroFile(AZStd::string filename)
+    {
+        AZ::IO::FileIOStream fileStream;
+        if (!fileStream.Open(filename.c_str(), AZ::IO::OpenMode::ModeRead | AZ::IO::OpenMode::ModeBinary))
+        {
+            return xacro::Params();
+        }
+
+        AZ::IO::SizeType length = fileStream.GetLength();
+        if (length == 0)
+        {
+            return xacro::Params();
+        }
+
+        AZStd::vector<char> charBuffer;
+        charBuffer.resize_no_construct(length + 1);
+        fileStream.Read(length, charBuffer.data());
+        charBuffer.back() = 0;
```

Comment:
nit: Here it might be clearer to use `'\0'` to indicate EOF character

Reply:
Or avoid the issue entirely and change the charBuffer type to an AZStd::string.
The `+1` on the length wouldn't be needed as the AZStd::string will set `charBuffer[length] = char{}` in  the [resize_no_construct](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/string/string.h#L1037-L1060) function

```
        AZStd::string charBuffer;
        charBuffer.resize_no_construct(length);
        fileStream.Read(length, charBuffer.data());
```

Or better yet, use the [resize_and_overwrite](https://en.cppreference.com/w/cpp/string/basic_string/resize_and_overwrite) function which will avoid the string from ever being in an invalid state

```suggestion
        AZStd::string charBuffer;
        auto ReadFile = [&fileStream](char* buffer, size_t size)
        {
            return systemFileStream.Read(size, buffer);
        };
        charBuffer.resize_and_overwrite(length, ReadFile);
```

Reply:
Unfortunately, rapid json needs a zero-terminated, mutable char array. I moved copy inside the parsing function.